### PR TITLE
`<MultiselectCheckbox/>`- using both builders and object

### PR DESCRIPTION
**Pull request by TalZa1810**
### 🔦 Summary
Supporting using builders in `<MultiselectCheckbox/>`. Supporting also the current component API for backwards compatibility.

### CODE REVIEW

**Comment by WixBuildServer** [comment]
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://www.wix.com/pages/wix-style-react.pr_5778/

File: src/ListItemSelect/ListItemSelect.js [hunk 1]
```diff
@@ -200,24 +200,28 @@ export const listItemSelectBuilder = ({
...
-      size={size}
-      ellipsis={ellipsis}
-      selected={selected}
-      highlighted={props.hovered}
-      {...props}
-    />
-  ),
-});
+}) => {
+  debugger;
```
Comment (lironch) [inline src/ListItemSelect/ListItemSelect.js:204]: oh no! 😱

**Comment by TalZa1810** [comment]
Few things I wasn't sure about:
1. API description
2. Story description.
3. Types changes (builders)

File: src/MultiSelectCheckbox/MultiSelectCheckbox.st.css [hunk 2]
```diff
@@ -0,0 +1,4 @@
+.showPointer,
```
Comment (shlomitc) [inline src/MultiSelectCheckbox/MultiSelectCheckbox.st.css:1]: please revert this. we decided to not do any further scss->st migration until v9

File: src/DropdownLayout/DropdownLayout.js [hunk 3]
```diff
@@ -412,7 +412,7 @@ class DropdownLayout extends WixComponent {
           selected: selected && selectedHighlight,
           hovered,
           disabled,
-          title,
+          title: title && !overrideStyle,
```
Comment (shlomitc) [inline src/DropdownLayout/DropdownLayout.js:415]: Why is this needed?
I think it's because you're misusing the `title` prop.
`DropdownLayout` expects `title` to define an item which is visual but isn't selectable.
`ListItemSelect` expects `title` to define the main title.

In that case, you need to change the builder so `title` will not be mapped to the DropdownLayout object

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.visual.js [hunk 4]
```diff
@@ -76,6 +76,32 @@ const interactiveTests = [
           valueParser: option => (option.label ? option.label : option.value),
         },
       },
+      {
+        it: 'Builders items',
```
Comment (TalZa1810) [inline src/MultiSelectCheckbox/test/MultiSelectCheckbox.visual.js:80]: I wasn't sure whether adding this visual test is necessary

File: src/MultiSelectCheckbox/MultiSelectCheckbox.js [hunk 5]
```diff
@@ -139,13 +150,17 @@ MultiSelectCheckbox.propTypes = {
 
   /** delimiter between the selected options that will be displayed in the input. */
   delimiter: PropTypes.string,
+
+  /** Function that receives an option, and should return the value to be displayed. By default returns `option.value` in case the option is an object and `option.optionTitle` in case it's a builder */
+  valueParser: PropTypes.func,
```
Comment (shlomitc) [inline src/MultiSelectCheckbox/MultiSelectCheckbox.js:155]: 1. I don't think that the component should not declare anything regarding the builder, so I'd remove it from the comment.

2. Regarding the `optionTitle` - now that's a bit more gentle case. It is an interesting idea to add it, but this becomes a public API of an option. The case where it becomes public is whenever the `option` is being sent as a callback to the consumer (like in `valueParser`)

I can suggest two options:
1. don't do this assistance. whoever decides uses `listItemBuilder` should supply it's own `valueParser`.
Since this `title` is basically in the hands of the consumer, he can simply take `id` which is already part of the required API, and combine it with the right title himself.
`<MultiSelectCheckbox valueParser(option => myOptionList[option.id].title)/>`

2. Define this API as part of a `DropdownLayout` option.
The `options` API of `DropdownLayout` which expects an array of `option`s will allow getting a new `label` property, same as it expects a `value` or `overrideStyles` here https://github.com/wix/wix-style-react/blob/9ed9ce86027fcd5c21ffe4e66330ce4ea9a80e13/src/DropdownLayout/DropdownLayout.js#L498
so basically `label` should always be the string applied in case of selection, unless `value` is a string` and then it can be used

Comment (TalZa1810) [reply src/MultiSelectCheckbox/MultiSelectCheckbox.js]: We saw that a lot of the consumers use anyways `label` in their own `valueParser` so maybe exposing `label` prop in the builder (instead of `optionTitle`) and have a default valueParser of `option => option.label || option.value` can be a solution (and updating the optionPropTypes to have a `label` as well of course).

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.visual.js (hunk 4 above)
Comment (shlomitc) [reply src/MultiSelectCheckbox/test/MultiSelectCheckbox.visual.js:80]: I don't think you need them

File: src/DropdownLayout/DropdownLayout.js [hunk 6]
```diff
@@ -501,6 +501,7 @@ const optionPropTypes = PropTypes.shape({
     .isRequired,
   disabled: PropTypes.bool,
   overrideStyle: PropTypes.bool,
+  label: PropTypes.string,
```
Comment (shlomitc) [inline src/DropdownLayout/DropdownLayout.js:504]: please add a test for it

File: src/InputWithOptions/InputWithOptions.js [hunk 7]
```diff
@@ -14,7 +14,7 @@ import uniqueId from 'lodash/uniqueId';
 
 import Popover from '../Popover';
 
-export const DEFAULT_VALUE_PARSER = option => option.value;
+export const DEFAULT_VALUE_PARSER = option => option.label || option.value;
```
Comment (shlomitc) [inline src/InputWithOptions/InputWithOptions.js:17]: and add a test for this

Comment (TalZa1810) [reply src/InputWithOptions/InputWithOptions.js]: This is used within the `MuiltiSelectCheckbox` itself and not in `InputWithOptions`. Moreover, it is used in `Dropdown` (since it extends `InputWithOption`.  Therefore, I've made a different `valueParser` specifically for `MultiSelectCheckbox` in order to to affect `Dropdown`.

File: src/MultiSelectCheckbox/docs/index.story.js [hunk 8]
```diff
@@ -77,9 +76,16 @@ export default {
...
-          code({
+          example({
+            title: 'Simple usage',
+            text: 'Simple usage of object options.',
             source: examples.simple,
-            compact: false,
+          }),
+
+          example({
+            title: 'Advanced usage',
```
Comment (shlomitc) [inline src/MultiSelectCheckbox/docs/index.story.js:86]: That's not an "Advanced Usage" but simple as it is - "Using builders".
And the description can be - "Use the Multi-selection dropdown behavior using the predefined list items.

In the future, when we create a well-documented page for the list items, this could be a good place to point to that doc

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js [hunk 9]
```diff
@@ -150,20 +150,82 @@ describe('multiSelectCheckbox', () => {
       );
     });
 
-    it('should use provided valueParser that will enable handling option with a component in value', async () => {
-      const specialOption = options.find(x => typeof x.value !== 'string');
-      const selectedOptions = [specialOption.id];
-      const valueParser = option =>
-        typeof option.value === 'string' ? option.value : option.label;
+    describe('valueParser', () => {
+      it('should use provided valueParser when given', async () => {
```
Comment (shlomitc) [inline src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:154]: Let's split this test into two:
1. it('should use the default valueParser') - this test will make sure that `value` is parsed.
2. it('should use the default valueParser for options contains `label` property) - this will make sure that `label` is parsed

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js [hunk 10]
```diff
@@ -150,20 +150,82 @@ describe('multiSelectCheckbox', () => {
...
+
+        const { driver, inputDriver } = createDriver(
+          <MultiSelectCheckbox
+            valueParser={valueParser}
+            options={options}
+            selectedOptions={selectedOptions}
+          />,
+        );
+        expect(await driver.getLabelAt(0)).toBe(specialOption.title);
+        expect(await inputDriver.getValue()).toBe(specialOption.title);
```
Comment (shlomitc) [inline src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:173]: Why is both `getLabelAt(0)` and `getValue()` is needed? looks to me like one of them is redundant. probably the second one

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js [hunk 11]
```diff
@@ -150,20 +150,82 @@ describe('multiSelectCheckbox', () => {
...
+        expect(await driver.getLabelAt(0)).toBe(options[0].label);
+        expect(await driver.getLabelAt(1)).toBe(options[1].value);
+
+        expect(await inputDriver.getValue()).toBe(
+          `${options[0].label}, ${options[1].value}`,
+        );
+      });
+    });
+
+    describe('Builder Options', () => {
```
Comment (shlomitc) [inline src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:205]: I wouldn't test it against a builder but rather against the new protocol, meaning `value` which is a function.

This makes sure the implementation is independent of any builder and supports the render method of an option

it should be very simple. It doesn't even have to have all the rendering options of the ListItemSelect.

Think that this test could have been written before any builder was created

File: src/Input/Input.js [hunk 12]
```diff
@@ -516,7 +516,7 @@ Input.propTypes = {
   /** Placement of status tooltips */
   tooltipPlacement: PropTypes.string,
 
-  /** Specifies the type of <input> element to display.default is text. */
+  /** Specifies the type of `<input/>` element to display.default is text. */
```
Comment (shlomitc) [inline src/Input/Input.js:519]: I don't think it needs to be mentioned as a component, it's referring an input field

File: src/DropdownLayout/DropdownLayout.js (hunk 6 above)
Comment (shlomitc) [reply src/DropdownLayout/DropdownLayout.js:504]: still missing a test here

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js (hunk 10 above)
Comment (TalZa1810) [reply src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:173]: The second one is referring to the value parser because it takes the value from the input. so I think that the first one is the one redundant.

File: src/Input/Input.js (hunk 12 above)
Comment (TalZa1810) [reply src/Input/Input.js:519]: It renders an input in the documentation:
https://www.wix.com/pages/wix-style-react/?activeTab=API&path=%2Fstory%2Fcomponents-api-components--input
I thought it was by mistake.

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js (hunk 10 above)
Comment (shlomitc) [reply src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:173]: From what I see in the driver - the first one encapsulates the second

File: src/DropdownLayout/DropdownLayout.js (hunk 6 above)
Comment (TalZa1810) [reply src/DropdownLayout/DropdownLayout.js:504]: I've added an if statement to the `optionValidator`. is this what you were referring?

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js [hunk 13]
```diff
@@ -150,20 +149,84 @@ describe('multiSelectCheckbox', () => {
...
+            valueParser={valueParser}
+            options={options}
+            selectedOptions={selectedOptions}
+          />,
+        );
+        expect(await driver.getLabelAt(0)).toBe(specialOption.title);
+      });
+    });
+
+    it('should allow using functions as options', async () => {
```
Comment (TalZa1810) [inline src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:211]: it this what you were referring?

File: src/DropdownLayout/DropdownLayout.js (hunk 6 above)
Comment (shlomitc) [reply src/DropdownLayout/DropdownLayout.js:504]: 1. it's missing in the prop types: https://github.com/wix/wix-style-react/blob/9ed9ce86027fcd5c21ffe4e66330ce4ea9a80e13/src/DropdownLayout/DropdownLayout.js#L498
2. add a test for the `DropdownLayout` that gets and item with a label and when clicking, the value is received in the callback https://github.com/wix/wix-style-react/blob/9ed9ce86027fcd5c21ffe4e66330ce4ea9a80e13/src/DropdownLayout/DropdownLayout.js#L420

File: src/MultiSelectCheckbox/docs/index.story.js (hunk 8 above)
Comment (shlomitc) [reply src/MultiSelectCheckbox/docs/index.story.js]: again, it's not "advanced"

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js (hunk 13 above)
Comment (shlomitc) [reply src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:211]: yes but the expectation should be testing the selection logic, meaning that you should have multiple actions clicked and each one's render function should be called with the "selected" being true.
https://github.com/wix/wix-style-react/pull/5778/files/eb368bc421f18f1a4c9f2ffaa5f0cdf52ce9ec1b#diff-fa476392e31f6bf3c05eb58029ab1267R19

File: src/MultiSelectCheckbox/docs/index.story.js (hunk 8 above)
Comment (TalZa1810) [reply src/MultiSelectCheckbox/docs/index.story.js]: I've changed this already: 
https://github.com/wix/wix-style-react/blob/eb368bc421f18f1a4c9f2ffaa5f0cdf52ce9ec1b/src/MultiSelectCheckbox/docs/index.story.js#L86

Or you refer also to the "Simple usage"?

Comment (shlomitc) [reply src/MultiSelectCheckbox/docs/index.story.js]: I don't see the changes reflected in this PR. maybe you didn't push it yet?

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js (hunk 11 above)
Comment (shlomitc) [reply src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js]: I see you're still using builders here - let me know if I need to reexplain myself here

File: src/MultiSelectCheckbox/docs/index.story.js (hunk 8 above)
Comment (TalZa1810) [reply src/MultiSelectCheckbox/docs/index.story.js]: Check out here: https://github.com/wix/wix-style-react/pull/5778/files

File: src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js (hunk 13 above)
Comment (TalZa1810) [reply src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js:211]: I've refactored this: 
https://github.com/wix/wix-style-react/blob/07b4809da663ab0f827402c9ce52bddeb69a4240/src/MultiSelectCheckbox/test/MultiSelectCheckbox.spec.js#L211
